types/time.go

```diff
@@ -1265,139 +1265,235 @@ func (d Duration) MicroSecond() int {
-				}
-				var dur Duration
-				dur, err1 = t.ConvertToDuration()
-				if err1 != nil {
-					return ZeroDuration, errors.Trace(err)
-				}
-				return dur.RoundFrac(fsp)
+			if i == 1 && requireColon {
+				return [3]int{}, str, err
 			}
+			break
 		}
-	case 2:
-		// HH:MM
-		_, err = fmt.Sscanf(integeralPart, "%2d:%2d", &hour, &minute)
-	case 3:
-		// Time format maybe HH:MM:SS or HHH:MM:SS.
-		// See https://dev.mysql.com/doc/refman/5.7/en/time.html
-		if len(seps[0]) == 3 {
-			_, err = fmt.Sscanf(integeralPart, "%3d:%2d:%2d", &hour, &minute, &second)
+	}
+
+	return hhmmss, rest, nil
+}
+
+func matchHHMMSSCompact(str string) ([3]int, string, error) {
+	num, rest, err := parser.Number(str)
+	if err != nil {
+		return [3]int{}, str, err
+	}
+	hhmmss := [3]int{num / 10000, (num / 100) % 100, num % 100}
+	return hhmmss, rest, nil
+}
+
+func hhmmssAddOverflow(hms []int, overflow bool) {
+	mod := []int{-1, 60, 60}
+	for i := 2; i >= 0 && overflow; i-- {
+		hms[i]++
+		if hms[i] == mod[i] {
+			overflow = true
+			hms[i] = 0
 		} else {
-			_, err = fmt.Sscanf(integeralPart, "%2d:%2d:%2d", &hour, &minute, &second)
+			overflow = false
 		}
-	default:
-		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", origStr)
+	}
+}
+
+func checkHHMMSS(hms [3]int) bool {
+	m, s := hms[1], hms[2]
+	return m < 60 && s < 60
+}
+
+// matchFrac returns overflow, fraction, rest, error
+func matchFrac(str string, fsp int8) (bool, int, string, error) {
+	rest, err := parser.Char(str, '.')
+	if err != nil {
+		return false, 0, str, nil
 	}
 
-	if terror.ErrorEqual(err, io.EOF) {
-		err = ErrTruncatedWrongVal.GenWithStackByArgs("time", origStr)
+	digits, rest, err := parser.Digit(rest, 0)
+	if err != nil {
+		return false, 0, str, err
+	}
+
+	frac, overflow, err := ParseFrac(digits, fsp)
+	if err != nil {
+		return false, 0, str, err
 	}
+
+	return overflow, frac, rest, nil
+}
+
+func matchDuration(str string, fsp int8) (Duration, error) {
+	fsp, err := CheckFsp(int(fsp))
 	if err != nil {
 		return ZeroDuration, errors.Trace(err)
 	}
 
+	if len(str) == 0 {
+		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
+	}
+
+	negative, rest := isNegativeDuration(str)
+	rest = parser.Space0(rest)
+
+	hhmmss := [3]int{}
+
+	if day, hms, remain, err := matchDayHHMMSS(rest); err == nil {
+		hms[0] += 24 * day
+		rest, hhmmss = remain, hms
+	} else if hms, remain, err := matchHHMMSSDelimited(rest, true); err == nil {
+		rest, hhmmss = remain, hms
+	} else if hms, remain, err := matchHHMMSSCompact(rest); err == nil {
+		rest, hhmmss = remain, hms
+	} else {
+		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
+	}
+
+	rest = parser.Space0(rest)
+	overflow, frac, rest, err := matchFrac(rest, fsp)
+	if err != nil || len(rest) > 0 {
+		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
+	}
+
 	if overflow {
-		second++
-		fracPart = 0
+		hhmmssAddOverflow(hhmmss[:], overflow)
```

Comment:
Why use `hhmmss[:]` here instead of `hhmmss` directly?

Reply:
The type of `hhmmss` is `[3]int`, an array of `int` instead of a slice